crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -444,6 +495,187 @@ impl Object {
+
+                    query
+                })
+            })
+            .await?;
+
+        let mut conn = Connection::new(prev, next);
+
+        for stored in results {
+            let cursor = stored.cursor().encode_cursor();
+            let object = Object::try_from(stored)?;
+            conn.edges.push(Edge::new(cursor, object));
+        }
+
+        Ok(conn)
+    }
+}
+
+impl ObjectFilter {
+    /// Try to create a filter whose results are the intersection of objects in `self`'s results and
+    /// objects in `other`'s results. This may not be possible if the resulting filter is
+    /// inconsistent in some way (e.g. a filter that requires one field to be two different values
+    /// simultaneously).
+    pub(crate) fn intersect(self, other: ObjectFilter) -> Option<Self> {
+        macro_rules! intersect {
+            ($field:ident, $body:expr) => {
+                intersect::field(self.$field, other.$field, $body)
+            };
+        }
+
+        // Treat `object_ids` and `object_keys` as a single filter on IDs, and optionally versions,
+        // and compute the intersection of that.
+        let keys = intersect::field(self.keys(), other.keys(), |k, l| {
+            let mut combined = BTreeMap::new();
+
+            for (id, v) in k {
+                if let Some(w) = l.get(&id).copied() {
+                    combined.insert(id, intersect::field(v, w, intersect::by_eq)?);
+                }
+            }
+
+            // If the intersection is empty, it means, there were some ID or Key filters in both
+            // `self` and `other`, but they don't overlap, so the final result is inconsistent.
+            (!combined.is_empty()).then_some(combined)
+        })?;
+
+        // Extract the ID and Key filters back out. At this point, we know that if there were ID/Key
+        // filters in both `self` and `other`, then they intersected to form a consistent set of
+        // constraints, so it is safe to interpret the lack of any ID/Key filters respectively as a
+        // lack of that kind of constraint, rather than a constraint on the empty set.
+
+        let object_ids = {
+            let partition: Vec<_> = keys
+                .iter()
+                .flatten()
+                .filter_map(|(id, v)| v.is_none().then_some(*id))
+                .collect();
+
+            (!partition.is_empty()).then_some(partition)
+        };
+
+        let object_keys = {
+            let partition: Vec<_> = keys
+                .iter()
+                .flatten()
+                .filter_map(|(id, v)| {
+                    Some(ObjectKey {
+                        object_id: *id,
+                        version: (*v)?,
+                    })
+                })
+                .collect();
+
+            (!partition.is_empty()).then_some(partition)
+        };
+
+        Some(Self {
+            type_: intersect!(type_, TypeFilter::intersect)?,
+            owner: intersect!(owner, intersect::by_eq)?,
```

Comment:
I think the missing piece here is that the intersection logic only applies by both `a` and `b` are not `None`, otherwise we pick whichever one isn't `None`.